### PR TITLE
Tests for genesis contracts

### PR DESCRIPTION
**Pull request by AyushKaul**
- [x] migrations
- [x] tests

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "testrpc": "ganache-cli --mnemonic 'clock radar mass judge dismiss just intact mind resemble fringe diary casino' --gasLimit 13000000",
+    "testrpc": "ganache-cli --hardfork istanbul -p 7545 --networkId 5777 --mnemonic 'clock radar mass judge dismiss just intact mind resemble fringe diary casino' --gasLimit 13000000 --gasPrice 0",
```
Comment (atvanguard) [inline package.json:7]: We use port 8545 for child chain, so let that be default.
Usually use the network id of 15001.
Remove `--gasPrice 0`.

Comment (AyushKaul) [reply package.json:7]: okay i'll change

**Comment by AyushKaul** [comment]
@atvanguard some functions are not used, I've marked them can you please tell me if I can remove them

File: package.json (hunk 1 above)
Comment (AyushKaul) [reply package.json]: done

File: contracts/System.sol [hunk 2]
```diff
@@ -4,7 +4,7 @@ contract System {
   address public constant SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;
 
   modifier onlySystem() {
-    require(msg.sender == SYSTEM_ADDRESS);
+    require(msg.sender == SYSTEM_ADDRESS, "Not System Addess!");
```
Comment (AyushKaul) [inline contracts/System.sol:7]: **contract change**

File: contracts/test/TestBorValidatorSet.sol [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+  function setSystemAddress(address _system) public {
+    dummySystem = _system;
+  }
+
+  modifier onlySystem() {
+    require(msg.sender == dummySystem);
+    _;
+  }
+}
```
Comment (AyushKaul) [inline contracts/test/TestBorValidatorSet.sol:17]: end of file

File: package.json [hunk 4]
```diff
@@ -4,11 +4,12 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "testrpc": "ganache-cli --mnemonic 'clock radar mass judge dismiss just intact mind resemble fringe diary casino' --gasLimit 13000000",
+    "testrpc": "ganache-cli -p 8545 --networkId 15001 --mnemonic 'clock radar mass judge dismiss just intact mind resemble fringe diary casino' --gasLimit 13000000",
```
Comment (AyushKaul) [inline package.json:7]: ?okay to fix the network id

File: test/BorValidatorSet.test.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+                currentSpan.number.add(new BN(1)),
+                currentSpan.endBlock.add(new BN(1)),
+                currentSpan.endBlock.add(new BN(256)),
+                validatorBytes,
+                producerBytes,
+                { from: accounts[0] }
+            )
+            // assert next span to be 1
+            const nextSpan = await testBVS.getNextSpan()
+            assertBigNumberEquality(nextSpan.number, new BN(1))
```
Comment (AyushKaul) [inline test/BorValidatorSet.test.js:87]: this fails

File: test/BorValidatorSet.test.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        })
+
+        it('Validator set for span #1', async () => {
+            const currentSpan = await testBVS.getCurrentSpan()
+            console.log("--------------------- for span 1 ------------------")
+            console.log(currentSpan)
+            const validators = await testBVS.getBorValidators(currentSpan.startBlock)
+            console.log("-------------------- validators ------------------")
+            console.log(validators)
+            assert.strictEqual(validators[0][0], accounts[0])
```
Comment (AyushKaul) [inline test/BorValidatorSet.test.js:119]: **fails here**

File: test/StateReceiver.test.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+const ethUtils = require('ethereumjs-util')
+const TestStateReceiver = artifacts.require('TestStateReceiver')
+const TestCommitState = artifacts.require('TestCommitState')
+const BN = ethUtils.BN
+
+contract('StateReceiver', async (accounts) => {
```
Comment (atvanguard) [inline test/StateReceiver.test.js:6]: Nice tests. Please try to test a couple more things.
- Calling `commitState` with a non-system address should revert.
- Make `onStateReceive` consume > 5M gas (that's the upper limit on gas we have hardcoded) and assert that `bool success` received from `commitState` is false. Assert true in other scenarios.
- One way to do this could be to decode `_data` in `onStateReceive` and run the loop that many times. To test the above scenario send in a large value so that loop runs out of gas.

File: package.json (hunk 4 above)
Comment (atvanguard) [reply package.json:7]: looks good.

File: contracts/BorValidatorSet.template [hunk 8]
```diff
@@ -349,4 +349,4 @@ contract BorValidatorSet is System {
   function innerNode(bytes32 left, bytes32 right) public pure returns (bytes32) {
     return sha256(abi.encodePacked(byte(uint8(1)), left, right));
   }
-}
+}
```
Comment (atvanguard) [inline contracts/BorValidatorSet.template:352]: Revert this change.

File: test/StateReceiver.test.js (hunk 7 above)
Comment (AyushKaul) [reply test/StateReceiver.test.js:7]: @atvanguard I've made the changes for the above. Is this what you wanted?
